sycl/doc/design/ThinLTO.md

```diff
@@ -0,0 +1,149 @@
+# ThinLTO for SYCL
+
+This document describes the purpose and design of ThinLTO for SYCL.
+
+**NOTE**: This is not the final version. The document is still in progress.
+
+## Background
+
+With traditional SYCL device code linking, all user code is linked together
+along with device libraries into a single huge module and then split and
+processed by `sycl-post-link`. This requires sequential processing, has a large
+memory footprint, and differs from the linking flow for AMD and NVIDIA devices.
+
+## Summary
+
+SYCL ThinLTO will hook into the existing community mechanism to run LTO as part
+of device linking inside `clang-linker-wrapper`. We split the device images
+early at compilation time, and at link time we use ThinLTO's function importing
+feature to bring in the definitions for referenced functions. Only the new
+offload model is supported.
+
+## Device code compilation time changes
+
+Most of the changes for ThinLTO occur during device link time, however there is
+one major change during compilation (-c) time: we now run device code split
+during compilation instead of linking. The main reason for doing this is
+increased parallelization. Many compilation jobs can be run at the same time,
+but linking happens once total for the application. Device code split is
+currently a common source of performance issues.
+
+Splitting early means that the resulting IR after splitting is not complete, it
```

Comment:
I don't think that early split leads to incomplete modules. Lack of linking leads to incomplete modules, i.e. a single translation unit with a call to undefined `SYCL_EXTERNAL` function is incomplete regardless of whether we split it or not.

If it was complete, then current version of the splitting mechanism won't break it, i.e. every resulting module would still be complete (at the cost of code duplication). However, I'm not 100% about the latter, if shared libraries are involved, that behavior definitely changes.

Reply:
sorry, you're right, let me reword it